consensus/obcpbft/pbft-core_test.go

```diff
@@ -1144,3 +1144,89 @@ func TestReplicaCrash2(t *testing.T) {
 		}
 	}
 }
+
+// TestReplicaCrash3 simulates the restart requiring a view change
+// to a checkpoint which was restored from the persistance state
+// Replicas 0,1,2 participate up to a checkpoint, then all crash
+// Then replicas 0,1,3 start back up, and a view change must be
+// triggered to get vp3 up to speed
+func TestReplicaCrash3(t *testing.T) {
+	validatorCount := 4
+	net := makePBFTNetwork(validatorCount, func(pep *pbftEndpoint) {
+		pep.pbft.K = 2
+		pep.pbft.L = 2 * pep.pbft.K
+	})
+	defer net.stop()
+
+	twoOffline := false
+	threeOffline := true
+	net.filterFn = func(src int, dst int, msg []byte) []byte {
+		if twoOffline && dst == 2 { // 2 is 'offline'
+			return nil
+		}
+		if threeOffline && dst == 3 { // 3 is 'offline'
+			return nil
+		}
+		return msg
+	}
+
+	mkreq := func(n int64) *Request {
+		txTime := &gp.Timestamp{Seconds: n, Nanos: 0}
+		tx := &pb.Transaction{Type: pb.Transaction_CHAINCODE_DEPLOY, Timestamp: txTime}
+		txPacked, _ := proto.Marshal(tx)
+
+		return &Request{
+			Timestamp: &gp.Timestamp{Seconds: n, Nanos: 0},
+			Payload:   txPacked,
+			ReplicaId: uint64(generateBroadcaster(validatorCount)),
+		}
+	}
+
+	for i := int64(1); i <= 8; i++ {
+		net.pbftEndpoints[0].pbft.recvRequest(mkreq(i))
+	}
+	net.process() // vp0,1,2 should have a stable checkpoint for seqNo 8
+
+	// Create new pbft instances to restore from persistence
+	for id := 0; id < 2; id++ {
+		pe := net.pbftEndpoints[id]
+		os.Setenv("CORE_PBFT_GENERAL_K", "2") // TODO, this is a hacky way to inject config before initialization rather than after, address this in a future changeset
```

Comment:
I agree, the problem is that the persistence recovery happens before I can modify the environment (and would set the low watermark incorrectly), hence the hack, really think we should be passing in a 'config struct' of some sort, rather than Viper (though we could use Viper to construct it), that was these sorts of hacks wouldn't be necessary.